// vega-transform-db/index.js
import { inherits, ingest, Transform } from "vega";
const querystring = require('querystring');
const http = require('http');

/**
 * Queries data from a Postgres database.
 * @constructor
 * @param {object} params - The parameters for this operator.
 */
export default function VegaTransformPostgres(params) {
  Transform.call(this, [], params);
}

VegaTransformPostgres.setHttpOptions = function (httpOptions) {
  if (httpOptions) {
    this._httpOptions = httpOptions;
    return this;
  }
  return this._httpOptions;
}

VegaTransformPostgres.Definition = {
  type: "dbtransform", // FixMe: make uppercase
  metadata: { changes: true, source: true },
  params: [
    { name: "relation", type: "string", required: true }
  ]
};

const prototype = inherits(VegaTransformPostgres, Transform);

prototype.transform = async function (params, pulse) {
  console.log(this)
  if (!VegaTransformPostgres._httpOptions) {
    throw Error("Vega Transform Postgres http options missing. Assign it with setHttpOptions.");
  }
  if (!this._query) {
    throw Error("Internal error: this._query should be defined");
  }

  //console.log(this.query(this._argval.ops[0]))

  let result = [];
  const postQuery = async function (query) {
    const options = VegaTransformPostgres._httpOptions
    const response = await fetch(options.url, {
      method: options.method,
      mode: options.mode,
      headers: options.headers,
      body: querystring.stringify({
        query: query
      })
    });

    // fetch won’t reject on HTTP error status even if the response
    // is an HTTP 404 or 500. Instead, it will resolve normally with
    // ok status set to false
    if (response.ok)
      return await response.json();
    else {
      // capture the error message
      const err = await response.json();
      throw Error(
        (err.error + ': ' + err.message).replace(/(\r\n|\n|\r)/gm, "")
      );
    }
  };

  try {

    this._sql = this._query(this, this._argval.from)
    console.log(this._sql)

    result = await postQuery(this._sql);
  } catch (error) {
    console.log(error);
  }

  result.forEach(ingest);
  const out = pulse.fork(pulse.NO_FIELDS & pulse.NO_SOURCE);

  out.rem = Array.isArray(this.value) ? this.value : [];
  console.log(out)
  if (this.__proto__.constructor.Definition.type === "Extent") {
    // changing the result format for Extent transform
    // [{min:10, max, 100}] -> [10, 100]
    const temp = [result[0].min, result[0].max]
    result = temp
  }
  this.value = out.add = out.source = result;


  return out;
}
